holoviews/core/options.py

```diff
@@ -806,21 +807,45 @@ def strongest_match(cls, overlay, mode):
 
 
     @classmethod
-    def collapse_element(cls, overlay, key=None, ranges=None, mode='data'):
+    def collapse_element(cls, overlay, ranges=None, mode='data', backend=None):
         """
         Finds any applicable compositor and applies it.
         """
-        from .overlay import Overlay
-        match = cls.strongest_match(overlay, mode)
-        if match is None: return overlay
-        (_, applicable_op, (start, stop)) = match
-        values = overlay.values()
-        sliced = Overlay.from_values(values[start:stop])
-        result = applicable_op.apply(sliced, ranges, key=key)
-        result = result.relabel(group=applicable_op.group)
-        output = Overlay.from_values(values[:start]+[result]+values[stop:])
-        output.id = overlay.id
-        return output
+        from .overlay import Overlay, CompositeOverlay
+        unpack = False
+        if not isinstance(overlay, CompositeOverlay):
+            overlay = Overlay([overlay])
+            unpack = True
+
+        prev_ids = tuple()
+        while True:
+            match = cls.strongest_match(overlay, mode)
+            if match is None:
+                if unpack and len(overlay) == 1:
+                    return overlay.values()[0]
+                return overlay
+            (_, applicable_op, (start, stop)) = match
+            if isinstance(overlay, Overlay):
+                values = overlay.values()
+                sliced = Overlay(values[start:stop])
+            else:
+                values = overlay.items()
+                sliced = overlay.clone(values[start:stop])
+            result = applicable_op.apply(sliced, ranges, backend)
+            if applicable_op.group:
+                result = result.relabel(group=applicable_op.group)
+            if isinstance(overlay, Overlay):
+                result = [result]
+            else:
+                result = list(zip(sliced.keys(), [result]))
+            overlay = overlay.clone(values[:start]+result+values[stop:])
+
+            # Guard against infinite recursion for no-ops
+            spec_fn = lambda x: not isinstance(x, CompositeOverlay)
+            new_ids = tuple(overlay.traverse(lambda x: id(x), [spec_fn]))
+            if new_ids == prev_ids:
+                return overlay
+            prev_ids = new_ids
```

Comment:
I am happy that we have ``StatisticalCompositorTest`` but I don't yet have a mental model of how this code extends how Compositor works.

Reply:
I'll quickly describe it:

a) Compositors can now be applied to individual elements not just overlays.
b) Compositors are applied iteratively until there are no more matches in the compositor definitions (needed to reduce Overlays with multiple elements that should be transformed)
c) If transfer_options is true, the options are transferred from the input element to the output element. Additionally plot options that also apply to the operation are transferred there.

Reply:
Ok. I think this is all stuff to demonstrate with examples when we finally expose ``Compositor`` as a useful thing to users.